torch/csrc/distributed/c10d/socket.cpp

```diff
@@ -0,0 +1,818 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include <c10d/socket.h>
+
+#include <cstring>
+#include <system_error>
+#include <thread>
+#include <utility>
+#include <vector>
+
+#ifdef _WIN32
+#include <mutex>
+
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#else
+#include <fcntl.h>
+#include <netdb.h>
+#include <netinet/tcp.h>
+#include <poll.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+#endif
+
+#include <fmt/format.h>
+
+#include <c10d/error.h>
+#include <c10d/exception.h>
+#include <c10d/logging.h>
+
+namespace c10d {
+namespace detail {
+namespace {
+#ifdef _WIN32
+
+// Since Winsock uses the name `WSAPoll` instead of `poll`, we alias it here
+// to avoid #ifdefs in the source code.
+const auto pollFd = ::WSAPoll;
+
+// Winsock's `getsockopt()` and `setsockopt()` functions expect option values to
+// be passed as `char*` instead of `void*`. We wrap them here to avoid redundant
+// casts in the source code.
+int getSocketOption(SOCKET s, int level, int optname, void* optval, int* optlen) {
+  return ::getsockopt(s, level, optname, static_cast<char*>(optval), optlen);
+}
+
+int setSocketOption(SOCKET s, int level, int optname, const void* optval, int optlen) {
+  return ::setsockopt(s, level, optname, static_cast<const char*>(optval), optlen);
+}
+
+// Winsock has its own error codes which differ from Berkeley's. Fortunately the
+// C++ Standard Library on Windows can map them to standard error codes.
+inline std::error_code getSocketError() noexcept {
+  return std::error_code{::WSAGetLastError(), std::system_category()};
+}
+
+inline void setSocketError(int val) noexcept {
+  ::WSASetLastError(val);
+}
+
+#else
+
+const auto pollFd = ::poll;
+
+const auto getSocketOption = ::getsockopt;
+const auto setSocketOption = ::setsockopt;
+
+inline std::error_code getSocketError() noexcept {
+  return lastError();
+}
+
+inline void setSocketError(int val) noexcept {
+  errno = val;
+}
+
+#endif
+
+// Suspends the current thread for one second.
+void delay() {
+  constexpr std::chrono::seconds pause{1};
+
+#ifdef _WIN32
+  std::this_thread::sleep_for(pause);
+#else
+  ::timespec req{};
+  req.tv_sec = pause.count();
+
+  // The C++ Standard does not specify whether `thread::sleep_for()` should be
+  // signal-aware; therefore, we use the `nanosleep()` syscall.
+  if (::nanosleep(&req, nullptr) != 0) {
+    std::error_code err = getSocketError();
+    // We don't care about error conditions other than EINTR since a failure
+    // here is not critical.
+    if (err == std::errc::interrupted) {
+      throw std::system_error{err};
+    }
+  }
+#endif
+}
+
+class SocketListenOp;
+class SocketConnectOp;
+} // namespace
+
+class SocketImpl {
+  friend class SocketListenOp;
+  friend class SocketConnectOp;
+
+ public:
+#ifdef _WIN32
```

Comment:
I don't have objections to the current solution, just wanna mention the past discussion on this. 

https://github.com/pytorch/pytorch/pull/47749 introduced a few windows-only files to minimize the number of `#ifdef` in the code. Looks like this PR is going towards the opposite direction. It's totally your call, especially given that you own this part of our code base. :)